### DISCUSSION TITLE
how to solve TypeError: 'module' object is not callable 01_pytorch_workflow

### DISCUSSION THREAD

**Discussion by wessamalhamidi** [mrdbourke/pytorch-deep-learning#228] [Q&A]
The code :

```python
def plot_perdictions(train_data = X_train,
                     train_labels = y_train,
                     test_data = X_test,
                     test_labels = y_test, 
                     predictions = None):
  plt.figure(figsize=(10, 7))
  
  # Plot training data in blue
  plt.scatter(train_data, train_labels, c='b', s=4, label = 'training data')

  # Plot test data in green
  plt.scatter(test_data, test_labels, c='g', s=4, label = 'testing data ')

  # Are there prediction
  if predictions is not None:
    # Plot the prediction if they exist
    plt.scatter(test_data, predictions, c='r', s=4, label= 'Predictions')

  # Show the legend
  plt.legend(prop ={'size': 14});
```

how can i solve this !!!

when i call the function it give me type error:

```
TypeError                                 Traceback (most recent call last)
[<ipython-input-38-9596213a7376>](https://localhost:8080/#) in <module>
----> 1 plot_perdictions();

[<ipython-input-37-d9af1e54b8e5>](https://localhost:8080/#) in plot_perdictions(train_data, train_labels, test_data, test_labels, predictions)
      4                      test_labels = y_test,
      5                      predictions = None):
----> 6   plt.figure(figsize=(10, 7))
      7 
      8   # Plot training data in blue


TypeError: 'module' object is not callable
```

**Comment by AxisMeetsWorld** [comment]
I am getting the same exact thing. I am wondering if something might have changed since the video was created, as I looked at the guide [here](https://www.learnpytorch.io/01_pytorch_workflow/) and it seems that we need to use `.figure.Figure()` as opposed to just `.figure()`. The same occurred with `plt.scatter()` as it only worked for me with `plt.pyplot.scatter()`. @mrdbourke can you please comment on whether this was the appropriate solution and whether we were doing anything wrong here? Thanks!

**Reply by mrdbourke** [reply]
Onto this now, will update this comment with what I find. Thank you for the heads up.

**Reply by mrdbourke** [reply]
Hmm, I'm not sure what's happening here.

I just ran [notebook 01](https://www.learnpytorch.io/01_pytorch_workflow/) end-to-end in Google Colab (unchanged) and it resulted in no errors.

Did you make sure to import matplotlib? (this might be the reason behind your `plt.pyplot.scatter()` method rather than `plt.scatter()`.

```python
import matplotlib.pyplot as plt
```

Where are you running the code?

Here is the function for reference:

```python
def plot_predictions(train_data=X_train, 
                     train_labels=y_train, 
                     test_data=X_test, 
                     test_labels=y_test, 
                     predictions=None):
  """
  Plots training data, test data and compares predictions.
  """
  plt.figure(figsize=(10, 7))

  # Plot training data in blue
  plt.scatter(train_data, train_labels, c="b", s=4, label="Training data")
  
  # Plot test data in green
  plt.scatter(test_data, test_labels, c="g", s=4, label="Testing data")

  if predictions is not None:
    # Plot the predictions in red (predictions were made on the test data)
    plt.scatter(test_data, predictions, c="r", s=4, label="Predictions")

  # Show the legend
  plt.legend(prop={"size": 14});
```

And the link to the notebook I just ran: https://colab.research.google.com/drive/1FNykJdzYR8K9uqnQJD_lQCcqmMIS5nS7?usp=sharing

**Reply by AxisMeetsWorld** [reply]
Odd that it worked one way for you, but differently for me. Could this be a versioning thing at all? I think I might have been running a version that was a little ahead of where we were at in the video. I am still a little perplexed by the `.Figure` call that was required. You were right on the `matplotlib.pyplot. as plt ` call. Here is what torch version I am getting:
2.0.1+cu118
Thanks so much for following up, and I do want to mention your course is great. Very thorough and at a solid pace. Learning a lot.

**Comment by mohfarama** [comment]
I have another issue. I typed the same code but it says " TypeError: plot_predictions() missing 4 required positional arguments: 'X_train', 'Y_train', 'X_test', and 'Y_test' "

Please help if you could

Here is My_code : 
import torch
from torch import nn
import matplotlib.pyplot as plt
torch.__version__
w = 0.7
b = 0.3
s=0  
e=1  
st = 0.02
X = torch.arange(s,e,st).unsqueeze(dim=1)
Y = w*X +b
len(X)
len(Y)
X
Y
''' splitting Data '''
train_split = int(0.8*len(X))
train_split
X_train = X[:40]
Y_train = Y[:40]
X_test = X[40:]
Y_test = Y[40:]
#print ("\n TSR_train is {} \n  FLA_train is {} \n".format(TSR_tr,FLA_tr))
''' Visualization '''
def plot_predictions(X_train,Y_train,X_test,Y_test,predictions=None):
  plt.figure(figsize=(10,7))
  plt.scatter(X_train,Y_train, c="b",s=4, label="Training dataX")
  plt.scatter(X_test,Y_test,c="g",s=4,label="Testing datax")
  if predictions is not None:
    plt.scatter(X_test,predictions,c="r",s=4,label="predictions")
  plt.legend(prop={"size":14})  ;
plot_predictions();
